runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchPortablePipelineTranslator.java

```diff
@@ -0,0 +1,601 @@
+    DataSet<WindowedValue<KV<K, V>>> inputDataSet = context.getDataSetOrThrow(inputPCollectionId);
+    RunnerApi.WindowingStrategy windowingStrategyProto =
+        pipeline
+            .getComponents()
+            .getWindowingStrategiesOrThrow(
+                pipeline
+                    .getComponents()
+                    .getPcollectionsOrThrow(inputPCollectionId)
+                    .getWindowingStrategyId());
+
+    RehydratedComponents rehydratedComponents =
+        RehydratedComponents.forComponents(pipeline.getComponents());
+
+    WindowingStrategy<Object, BoundedWindow> windowingStrategy;
+    try {
+      windowingStrategy =
+          (WindowingStrategy<Object, BoundedWindow>)
+              WindowingStrategyTranslation.fromProto(windowingStrategyProto, rehydratedComponents);
+    } catch (InvalidProtocolBufferException e) {
+      throw new IllegalStateException(
+          String.format(
+              "Unable to hydrate GroupByKey windowing strategy %s.", windowingStrategyProto),
+          e);
+    }
+
+    WindowedValueCoder<KV<K, V>> inputCoder =
+        instantiateCoder(inputPCollectionId, pipeline.getComponents());
+
+    KvCoder<K, V> inputElementCoder = (KvCoder<K, V>) inputCoder.getValueCoder();
+
+    Concatenate combineFn = new Concatenate<>();
+    Coder<List<V>> accumulatorCoder =
+        combineFn.getAccumulatorCoder(
+            CoderRegistry.createDefault(), inputElementCoder.getValueCoder());
+
+    Coder<WindowedValue<KV<K, List<V>>>> outputCoder =
+        WindowedValue.getFullCoder(
+            KvCoder.of(inputElementCoder.getKeyCoder(), accumulatorCoder),
+            windowingStrategy.getWindowFn().windowCoder());
+
+    TypeInformation<WindowedValue<KV<K, List<V>>>> partialReduceTypeInfo =
+        new CoderTypeInformation<>(outputCoder);
+
+    Grouping<WindowedValue<KV<K, V>>> inputGrouping =
+        inputDataSet.groupBy(new KvKeySelector<>(inputElementCoder.getKeyCoder()));
+
+    FlinkPartialReduceFunction<K, V, List<V>, ?> partialReduceFunction =
+        new FlinkPartialReduceFunction<>(
+            combineFn, windowingStrategy, Collections.emptyMap(), context.getPipelineOptions());
+
+    FlinkReduceFunction<K, List<V>, List<V>, ?> reduceFunction =
+        new FlinkReduceFunction<>(
+            combineFn, windowingStrategy, Collections.emptyMap(), context.getPipelineOptions());
+
+    // Partially GroupReduce the values into the intermediate format AccumT (combine)
+    GroupCombineOperator<WindowedValue<KV<K, V>>, WindowedValue<KV<K, List<V>>>> groupCombine =
+        new GroupCombineOperator<>(
+            inputGrouping,
+            partialReduceTypeInfo,
+            partialReduceFunction,
+            "GroupCombine: " + pTransform.getUniqueName());
+
+    Grouping<WindowedValue<KV<K, List<V>>>> intermediateGrouping =
+        groupCombine.groupBy(new KvKeySelector<>(inputElementCoder.getKeyCoder()));
+
+    // Fully reduce the values and create output format VO
```

Comment:
<!--new_thread; commit:c3bc9f1ff4d5f830b19aef83b758b12e363f7679; resolved:0-->
Why do we use partial reduction just for a GroupByKey? This looks more like translation for Combine.

Reply:
It looks like the flink sequence is a `DataSet` of `KV`s gets turned into a `Grouping` of `KV`s, but beam requires a `K, Iter<V>` for each `K`, and this is the way to produce that single element per key - so we need to perform the post-shuffle combine. The precombine is just a way to reduce the number of times we shuffle the key. We can probably do something clever to not force the entire dataset to fit in memory by producing filtered concatenated iterables of input elements, but I'd separate that change out; we might want to perform some of the precombine anyway to force a precondition on our inputs, but I'm not entirely sure about the specifics of Flink's GroupByKey/AndByWindow implementations

Filed [BEAM-4228](https://issues.apache.org/jira/browse/BEAM-4228) for this (which is quite possibly a dupe, but I haven't checked)